broadinstitute/pearl
# JN-552 published version names

This adds "published_version" to all versioned entities.  This is a 'cosmetic' version intended to help make the published dataset cleaner.  The UI display of the published version is very rough, but that can be refined later, the important thing for this PR is to get the data infrastructure in place.  For now, the published version is auto-generated.  The critical thing is to never rely on the published version for application logic.  That will ensure that published and unpublished surveys/content behave the same, and therefore the sandboxes will accurately mirror production.

Along the way, I upgraded our publishing services to make them more typesafe and add proper integration tests in place of the prior mocked-out test. 

TO REVIEW:
1. The real action is the `assignPublishedVersionIfNeeded` methods in PublishingUtils.java, and how they are called from StudyPublishingService and PortalPublishingService.  Start with those methods and move outwards.
2. The majority of this PR is a bunch of boring refactoring (adding generics, pulling out a base class, extracting utility methods, etc...) that can be skimmed. 

TO TEST:
1. repopulate OurHealth
4. Go to https://localhost:3000/ourhealth/studies/ourheart/publishing 
5.  copy the sandbox environment to the irb environment.  
6. copy the irb environment to the live environment
7. go to https://localhost:3000/ourhealth/studies/ourheart/env/sandbox/forms/preEnroll/oh_oh_preenroll?readOnly=false
8. Confirm the survey now shows a published version
<img width="984" alt="image" src="https://github.com/broadinstitute/juniper/assets/2800795/76dfb120-56f7-4973-9697-46c8cdd77d31">
